### PR TITLE
Fix asserts in destructors

### PR DESCRIPTION
**Pull request by masterleinad**
This is again resulting from the Coverty scan.
Calling `Assert` or `AssertNothrow` is uncritical as long as `disable_abort_on_exception` was not called. By wrapping possibly offending functions in a `try-catch-block`, we avoid double-exception situations and hiding exceptions also in the latter case.

Even if we only call `AssertNothrow`, Coverty still thinks that we might `throw`. Hence, there are also some false-positives in the report of the scan.

### CODE REVIEW

File: include/deal.II/lac/block_matrix_base.h [hunk 1]
```diff
@@ -1495,7 +1495,12 @@ template <typename MatrixType>
 inline
 BlockMatrixBase<MatrixType>::~BlockMatrixBase ()
 {
-  clear ();
+  try
+    {
+      clear ();
+    }
+  catch (...)
+    {}
```
Comment (tamiko) [inline include/deal.II/lac/block_matrix_base.h]: An `Assert` here is clearly a problem.
On the other hand, `AssertNothrow` does never throw. I implemented this macro exactly because of asserts in destructors. So can you please check whether, catching and issuing `AssertNothrow` works?
```
try { ...} catch (...) {AssertNothrow ...}
```

File: source/lac/trilinos_vector_base.cc [hunk 2]
```diff
@@ -372,7 +372,7 @@ namespace TrilinosWrappers
     // is zero on _all_ processors.
     const Epetra_MpiComm *mpi_comm
       = dynamic_cast<const Epetra_MpiComm *>(&vector->Map().Comm());
-    Assert(mpi_comm != 0, ExcInternalError());
+    AssertThrow(mpi_comm != 0, ExcInternalError());
```
Comment (tamiko) [inline source/lac/trilinos_vector_base.cc]: Why did you change all the `Assert` to `AssertThrow`. We have the convention that internal errors (such as this one) should be governed by `Assert`, not `AssertThrow`. (By the way, both can throw an exception - the first one is meant for internal error conditions, the latter is meant for error conditions due to user input/interaction that should always be thrown).

Comment (masterleinad) [inline source/lac/trilinos_vector_base.cc]: The point here was that we don't check whether the `dynamic_cast` fails in a Release build, if we only use `Assert`.

File: include/deal.II/lac/block_matrix_base.h (hunk 1 above)
Comment (masterleinad) [inline include/deal.II/lac/block_matrix_base.h]:
```
try {Assert/AssertThrow/AssertNothrow ( false, ExcMessage("3"));}
  catch (...) {AssertNothrow(false, ExcMessage("4"));}
```

results for `dealii::deal_II_exceptions::disable_abort_on_exception();` in

-  AssertThrow:    DEAL::Exception: ExcMessage("4"), no abort
-  Assert:              DEAL::Exception: ExcMessage("4"), no abort
-  AssertNothrow: DEAL::Exception: ExcMessage("3"), no abort

and without `dealii::deal_II_exceptions::disable_abort_on_exception();` in
-  AssertThrow:    Abort due to `AssertNothrow(false, ExcMessage("4"))`
-  Assert:              Abort due to `AssertNothrow(false, ExcMessage("3"))`
-  AssertNothrow: Abort due to `AssertNothrow(false, ExcMessage("3"))`

This behavior does not change when wrapping everything in another `try{...} catch(...) {}` block.
In the end, this is exactly what I would expect. Does this answer your question?

File: source/lac/trilinos_vector_base.cc (hunk 2 above)
Comment (tamiko) [inline source/lac/trilinos_vector_base.cc]: Other question - do we need a dynamic_cast at that point at all? After all the type should be known at compile time.

Replacing the dynamic_cast with a static_cast would solve both problems, (a) no runtime error condition whether the cast succeeds and (b) no `AssertThrow` (which is the wrong macro for an internal error).

File: source/base/timer.cc [hunk 3]
```diff
@@ -353,8 +353,13 @@ TimerOutput::TimerOutput (MPI_Comm      mpi_communicator,
 
 TimerOutput::~TimerOutput()
 {
-  while (active_sections.size() > 0)
-    leave_subsection();
+  try
+    {
+      while (active_sections.size() > 0)
+        leave_subsection();
```
Comment (bangerth) [inline source/base/timer.cc]: What is it actually that the checker complains about here? That `leave_subsection()` is not marked as `nothrow`? I'm confused what exactly it is that we can't do in constructors. I know that we can't throw exceptions, but it looks like the checker marks more or less random functions we happen to call as potentially throwing exceptions. Would it be better to instead mark these functions as `nothrow`?

File: source/lac/trilinos_vector_base.cc (hunk 2 above)
Comment (masterleinad) [inline source/lac/trilinos_vector_base.cc]: I think we should at least test in a Debug build that the cast is successful. It should be safe to say that the type of that object does not change from Debug to Release. Hence I am now convinced that an `Assert` is the right thing to do.

File: source/base/timer.cc (hunk 3 above)
Comment (masterleinad) [inline source/base/timer.cc]: It complains about three exceptions:
- `dealii::ExceptionBase` from `Assert(!active_sections.empty(),...)` in line 413
- `dealii::StandardExceptions::ExcMPI` from an `AssertThrowMPI`in `source/base/mpi.cc:225` (we call `Timer::stop` which calls `Utilities::MPI::min_max_avg`)
- `dealii::StandardExceptions::ExcMessage` from `Assert(sections.find (section_name) != sections.end(),...)`in line 420

Comment (bangerth) [inline source/base/timer.cc]: Hm, I'm not terribly excited about wrapping content of destructors into try-catch clauses just because some function can throw :-( But then I also don't have a good solution otherwise...

Comment (bangerth) [inline source/base/timer.cc]: Can we at least terminate the program, rather than ignoring the exception in the catch clause? It would seem to me that nothing positive can come out of continuing here.

**Comment by tamiko** [comment]
@masterleinad *umh* Wasn't there a short discussion about `dynamic_cast` versus `static_cast` on this pull request yesterday?

I thought I made multiple comments on this bug. Now I can't find them any more.

**Comment by bangerth** [comment]
It's in one of the out-dated comments above. You can still see it if you click on it -- it expands and shows you the discussion.

**Comment by masterleinad** [comment]
Yes, I changed back to `Assert` instead of `AssertThrow` realizing that this should indeed be sufficient. Afterwards you commented again, so this discussion is hidden.

File: source/base/timer.cc (hunk 3 above)
Comment (masterleinad) [inline source/base/timer.cc]: Then we would not be able to catch any exceptions that triggered the destruction of this object. What about abort if we catch `std::exception &` and otherwise ignore a thrown exception?

Comment (bangerth) [inline source/base/timer.cc]: Hm, you're right about stack unwinding upon an exception. I guess we should just go with your patch then...

**Comment by masterleinad** [comment]
Are there any open issues I should adress? More opinions on changing the `dynamic_cast` to a `static_cast` in 

```
    const Epetra_MpiComm *mpi_comm
       = dynamic_cast<const Epetra_MpiComm *>(&vector->Map().Comm());
    Assert(mpi_comm != 0, ExcInternalError());
```
?

**Comment by kronbichler** [comment]
It should indeed be possible to do a `static_cast` to `Epetra_MpiComm` because that is the only possibility when we are using MPI. Why not use the `Assert` with the dynamic cast and then the `static_cast` for actually using it? We use this idiom in several places in the library, e.g. line 2579 of `mapping_q_generic.cc`.

**Comment by masterleinad** [comment]
@kronbichler That sounds good. I'll create a separate PR for this.

**Comment by masterleinad** [comment]
@kronbichler It turns out that ` Epetra_MpiComm` is virtually inherited from `Epetra_Comm`. Hence, we can only use `dynamic_cast` and not `static_cast`.

**Comment by kronbichler** [comment]
Alright, then we just leave it at the dynamic cast and the subsequent `Assert`.

**Comment by tamiko** [comment]
Well, well. We can revisit all these places later when we switched to C++11.
At least clang and gcc implement a `noexcept` annotated function to always terminate instead of throwing an exception (that would unwind the function call).

**Comment by bangerth** [comment]
Yes, `noexcept` is the standard way to do this in C++11. Before that, it was already possible to annotate functions with `throw()`. As a matter of fact, we already have `DEAL_II_NOEXCEPT` that we could use to annotate some functions.
